components/script/dom/indexeddb/idbtransaction.rs

```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
+        // Abort can be initiated any time before the transaction finishes.
+        // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+        if self.finished.get() || self.aborted.get() {
+            return;
+        }
+
+        if self.error.get().is_none() {
+            if let Ok(exception) = create_dom_exception(&self.global(), error, can_gc) {
+                self.error.set(Some(&exception));
+            }
+        }
+
+        self.aborted.set(true);
+        let _ = self
+            .get_idb_thread()
+            .send(IndexedDBThreadMsg::AbortTransaction {
+                origin: self.global().origin().immutable().clone(),
+                name: self.db.get_name().to_string(),
+                transaction: self.serial_number,
+            });
+        // https://w3c.github.io/IndexedDB/#abort-a-transaction
+        // When a transaction is aborted the implementation must undo (roll back)
+        // any changes that were made to the database during that transaction.”
+        self.queue_abort_pending_requests();
+        // https://w3c.github.io/IndexedDB/#abort-a-transaction
+        if let Some(old_version) = self.upgrade_old_version.get() {
+            // Keep the DOM-side version consistent until backend cleanup runs.
+            self.db.set_version_cache(old_version);
+        }
+
+        // NOTE: Completion of abort waits for outstanding requests to finish.
         if self.pending_request_count.get() == 0 {
+            self.maybe_finish_after_abort();
+        }
+    }
+
+    fn queue_abort_pending_requests(&self) {
+        // https://w3c.github.io/IndexedDB/#transaction-concept
+        // A transaction has a request list of pending requests which have been made against the transaction.
+        // When a transaction is aborted the implementation must
+        // abort all outstanding requests made against the transaction.
+        let requests: Vec<Trusted<IDBRequest>> = self
+            .requests
+            .borrow()
+            .iter()
+            .map(|request| Trusted::new(&**request))
+            .collect();
+        let global = self.global();
+        global.task_manager().database_access_task_source().queue(
+            task!(abort_pending_requests: move || {
+                for request in requests {
+                    let request = request.root();
+                    request.abort_due_to_transaction(CanGc::note());
+                }
+            }),
+        );
+    }
+
+    fn inc_pending_requests(&self) {
+        // Track outstanding requests so we can detect when the transaction may finish.
+        self.pending_request_count
+            .set(self.pending_request_count.get() + 1);
+    }
+
+    fn dec_pending_requests(&self) {
+        let pending = self.pending_request_count.get();
+        if pending == 0 {
+            return;
+        }
+        self.pending_request_count.set(pending - 1);
+    }
+
+    fn maybe_finish_successfully(&self) {
```

Comment:
Again I think you first need to wait on the backend for all requests to have completed, and then communicate back to script to continue the finish logic here.